velox/docs/develop/types.rst

```diff
@@ -2,54 +2,107 @@
-======================  ===========================    ==================
-BOOLEAN                 bool                            0.125 (i.e. 1 bit)
-TINYINT                 int8_t                          1
-SMALLINT                int16_t                         2
-INTEGER                 int32_t	                        4
-BIGINT                  int64_t                         8
-DATE                    struct Date                     8
-REAL                    float                           4
-DOUBLE                  double                          8
-SHORT_DECIMAL           struct UnscaledShortDecimal     8
-LONG_DECIMAL            struct UnscaledLongDecimal     16
-TIMESTAMP               struct Timestamp               16
-INTERVAL DAY TO SECOND  struct IntervalDayTime          8
-VARCHAR                 struct StringView              16
-VARBINARY               struct StringView              16
-======================  ===========================    ==================
-
-All scalar types except DECIMAL have a one-to-one mapping with their C++ types.
-
-DECIMAL type is a special scalar type since it carries additional `precision`,
-and `scale` information. Similar to SQL decimal, `precision` is the number of
+Physical Types
+~~~~~~~~~~~~~~
+Each physical type is implemented using a C++ type. The table
+below shows the supported physical types, their corresponding C++ type,
+and bytes required per value.
+
+================   ===========================   ===================
+Physical Type      C++ Type                      Bytes per Value
+================   ===========================   ===================
+BOOLEAN            bool                          0.125 (i.e. 1 bit)
+TINYINT            int8_t                        1
+SMALLINT           int16_t                       2
+INTEGER            int32_t	                     4
+BIGINT             int64_t                       8
+HUGEINT            int128_t                      16
+DATE               struct Date                   8
+REAL               float                         4
+DOUBLE             double                        8
+TIMESTAMP          struct Timestamp              16
+VARCHAR            struct StringView             16
+VARBINARY          struct StringView             16
+OPAQUE             std::shared_ptr<void>         User-defined
+UNKNOWN            struct UnknownValue           0
+================   ===========================   ===================
+
+All physical types have a one-to-one mapping with their C++ types.
+The C++ type is also used as a template parameter for vector classes.
+
+StringView is a struct that contains a 4-byte size field, a 4-byte prefix field,
+and an 8-byte field to store the remaining string (after the prefix) in-line or
+a pointer to the full string out-of-line.
+Refer :doc:`vectors </develop/vectors>` for how vectors of StringView are stored.
+
+Velox allows users to define custom types by extending the OPAQUE type.
```

Comment:
Using Opaque as a base for a custom type is an option, but not the only option. Custom types are just logical types defined by the application. I would not encourage users to use OPAQUE type. Let's reword accordingly.

Reply:
Very good point! I think most custom types are logical types extending physical types. OPAQUE types must be used only when there is no physical type available to back a logical type.